### PR TITLE
SharedPtr improvements

### PR DESCRIPTION
**Pull request by glebm**
1. Moves all logic from `SharedImpl` to `SharedObj` (previously split between the two).
2. Moves all methods into the header so that the compiler can inline them.
3. Adds `operator=(T*)` to avoid the overhead of a temporary `SharedImpl` (made possible by #2752).
4. Adds C++ unit tests.
5. Removes useless methods: move constructors, value constructors, and move assignments.
6. Removes `virtual` from the `SharedPtr` destructor.

Formatted and laid out for readability and consistency.

SharedPtr methods are no longer at the top of the profile. Compiling Bootstrap 4:

```
Showing nodes accounting for 249156545, 78.41% of 317756978 total
Dropped 62 nodes (cum <= 1588784)
      flat  flat%   sum%        cum   cum%
  25726721  8.10%  8.10%   25726721  8.10%  [[kernel.kallsyms]]
  16355966  5.15% 13.24%   16355966  5.15%  [libc-2.27.so]
  14332598  4.51% 17.75%   14332598  4.51%  __libc_malloc
  12009839  3.78% 21.53%   12009839  3.78%  std::vector::_M_realloc_insert
   9173948  2.89% 24.42%   10269653  3.23%  _int_free (inline)
   8770958  2.76% 27.18%    8770958  2.76%  Sass::Cast
   8247287  2.60% 29.78%    8247287  2.60%  std::__cxx11::basic_string::_M_construct
   6533856  2.06% 31.83%    6533856  2.06%  Sass::Eval::operator()
   5700291  1.79% 33.63%    5700291  1.79%  Sass::Prelexer::spaces
   5481519  1.73% 35.35%    5481519  1.73%  _init
   5445579  1.71% 37.07%   15715232  4.95%  cfree
   5410714  1.70% 38.77%    5410714  1.70%  Sass::Prelexer::identifier
   5136915  1.62% 40.39%    5136915  1.62%  _int_malloc
   5030198  1.58% 41.97%    5030198  1.58%  Sass::SharedPtr::decRefCount
   4586955  1.44% 43.41%    4586955  1.44%  Sass::Prelexer::unicode_seq
   4476814  1.41% 44.82%    4476814  1.41%  Sass::Parser::parse_factor
   4464953  1.41% 46.23%    4464953  1.41%  operator new
   4391735  1.38% 47.61%    4391735  1.38%  Sass::Inspect::operator()
   4385350  1.38% 48.99%    4385350  1.38%  __dynamic_cast
   4365123  1.37% 50.36%    4365123  1.37%  __cxxabiv1::__si_class_type_info::__do_dyncast
   4114112  1.29% 51.66%    4114112  1.29%  do_lookup_x
   4111904  1.29% 52.95%    4111904  1.29%  Sass::Prelexer::identifier_alpha
   3292625  1.04% 53.99%    3292625  1.04%  Sass::Extend::extendSelectorList
   3291827  1.04% 55.02%    3291827  1.04%  Sass::Complex_Selector::hash
   3291774  1.04% 56.06%    3291774  1.04%  Sass::Selector_List::~Selector_List
   3289646  1.04% 57.09%    3289646  1.04%  __strtof_l
   3289553  1.04% 58.13%    3289553  1.04%  Sass::Compound_Selector::~Compound_Selector
   3284860  1.03% 59.16%    3284860  1.03%  Sass::Expand::operator()
   3284272  1.03% 60.20%    3284272  1.03%  Sass::Prelexer::alternatives
   3273931  1.03% 61.23%    3273931  1.03%  Sass::Environment::has_local
   3240576  1.02% 62.25%    3240576  1.02%  Sass::Offset::Offset
   2852001   0.9% 63.14%    2852001   0.9%  Sass::Prelexer::css_comments
   2621269  0.82% 63.97%    2621269  0.82%  Sass::Prelexer::end_of_line
   2250432  0.71% 64.68%    2250432  0.71%  Sass::Prelexer::alpha
   2208632   0.7% 65.37%    2208632   0.7%  Sass::Prelexer::sequence
   2195717  0.69% 66.06%    2195717  0.69%  Sass::Inspect::Inspect
   2195705  0.69% 66.75%    2195705  0.69%  Sass::Offset::add
   2195653  0.69% 67.45%    2195653  0.69%  Sass::Compound_Selector::is_superselector_of
   2195580  0.69% 68.14%    2195580  0.69%  Sass::Remove_Placeholders::remove_placeholders
   2194889  0.69% 68.83%    2194889  0.69%  Sass::Emitter::append_string
   2194455  0.69% 69.52%    2194455  0.69%  Sass::SourceMap::append
   2194336  0.69% 70.21%    2194336  0.69%  Sass::string_to_output
   2194245  0.69% 70.90%    2194245  0.69%  malloc_consolidate
   2193623  0.69% 71.59%    2193623  0.69%  std::vector::emplace_back
   2191941  0.69% 72.28%    2191941  0.69%  __GI___printf_fp_l
   2190831  0.69% 72.97%    2190831  0.69%  __exp1_fma
   2190529  0.69% 73.66%    2190529  0.69%  Sass::Complex_Selector::resolve_parent_refs
   2190320  0.69% 74.35%    2190320  0.69%  Sass::AST_Node::to_string[abi:cxx11]
   2189952  0.69% 75.04%    2189952  0.69%  Sass::Environment::find_local
   2189855  0.69% 75.73%    2189855  0.69%  Sass::bind
   2167116  0.68% 76.41%    2167116  0.68%  Sass::Prelexer::digits
   2147078  0.68% 77.08%    2147078  0.68%  Sass::Parser::parse_block_node
   2110929  0.66% 77.75%    2110929  0.66%  Sass::Prelexer::line_comment
   2109058  0.66% 78.41%    2109058  0.66%  Sass::CheckNesting::visit_children
```

### CODE REVIEW

File: test/Makefile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+CXX ?= c++
```
Comment (xzyfer) [inline test/Makefile]: It'd be good to update Travis and appveyor configuration to run these tests in CI also

Comment (glebm) [reply test/Makefile]: Done (for AUTOTOOLS=no only)

**Comment by glebm** [comment]
Based on Travis run time, this is a significant performance win.

Can you release a new version once this is merged?

All the memory leaks reported by ASan have been fixed, and I am not any closer to figuring out / reproducing the failing tests in the ASan build but pretty sure they are not bugs but issues with the test harness.

**Comment by glebm** [comment]
Rebased

**Pull request merged**